sphinx_doc/tutorials/docs/writing_new_costmap2d_plugin.rst

```diff
@@ -0,0 +1,195 @@
+
+2- Enable the plugin in Costmap2D
+---------------------------------
+
+1. At first, plugin's class should be linked with parent in ``costmap_plugins.xml``:
+
+.. code-block:: diff
+
+  @@ -12,6 +12,9 @@
+       <class type="nav2_costmap_2d::VoxelLayer"     base_class_type="nav2_costmap_2d::Layer">
+         <description>Similar to obstacle costmap, but uses 3D voxel grid to store data.</description>
+       </class>
+  +    <class type="nav2_costmap_2d::GradientLayer"  base_class_type="nav2_costmap_2d::Layer">
+  +      <description>Gradient layer for testing purposes.</description>
+  +    </class>
+     </library>
+   </class_libraries>
+
+2. On next step the source code of ``GradientLayer`` should be added to ``CMakeLists.txt`` in order to be included into final shared library:
+
+.. code-block:: diff
+
+  @@ -83,6 +83,7 @@ add_library(layers SHARED
+     plugins/obstacle_layer.cpp
+     src/observation_buffer.cpp
+     plugins/voxel_layer.cpp
+  +  plugins/gradient_layer.cpp
+   )
+   ament_target_dependencies(layers
+     ${dependencies}
+
+3. Finally, new plugin should be added to ``plugin_names`` and ``plugin_types`` variables in ``nav2_params.yaml`` optionally for ``local_costmap``/``global_costmap`` in order to be enabled in run-time for Controller/Planner Server. For example:
+
+.. code-block:: diff
+
+  @@ -130,8 +130,8 @@ local_costmap:
+         width: 3
+         height: 3
+         resolution: 0.05
+  -      plugin_names: ["obstacle_layer", "voxel_layer", "inflation_layer"]
+  -      plugin_types: ["nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::VoxelLayer", "nav2_costmap_2d::InflationLayer"]
+  +      plugin_names: ["obstacle_layer", "voxel_layer", "gradient_layer"]
+  +      plugin_types: ["nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::VoxelLayer", "nav2_costmap_2d::GradientLayer"]
+         robot_radius: 0.22
+         inflation_layer:
+           cost_scaling_factor: 3.0
+  @@ -177,8 +177,8 @@ global_costmap:
+         robot_base_frame: base_link
+         global_frame: map
+         use_sim_time: True
+  -      plugin_names: ["static_layer", "obstacle_layer", "voxel_layer", "inflation_layer"]
+  -      plugin_types: ["nav2_costmap_2d::StaticLayer", "nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::VoxelLayer", "nav2_costmap_2d::InflationLayer"]
+  +      plugin_names: ["static_layer", "obstacle_layer", "voxel_layer", "gradient_layer"]
+  +      plugin_types: ["nav2_costmap_2d::StaticLayer", "nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::VoxelLayer", "nav2_costmap_2d::GradientLayer"]
+         robot_radius: 0.22
+         resolution: 0.05
+         obstacle_layer:
+
+3- Re-build Costmap2D
+---------------------
+
+Re-build ``nav2_costmap_2d`` package with all plugins onboard (including ``GradientLayer``):
+
+.. code-block:: bash
+
+  $ cd ~/navigation2_ws
+  $ colcon build --packages-select nav2_costmap_2d --symlink-install
```

Comment:
If this is a separate package, you should not need to rebuild nav2_costmap_2d to use it